### PR TITLE
Proper parsing of elements and attribute dicts

### PR DESCRIPTION
**Pull request by rowanseymour**
Pulls in older work by @bcoughlan and @a1s, adds support for boolean attributes and Python 3

### CODE REVIEW

**Comment by rowanseymour** [comment]
Agree re code structure. Is that something you want to tackle?

We should also keep an eye on performance. From just playing with the tests I noticed a very minimal slow down with these changes, but worth it for proper parsing.

**Comment by Psycojoker** [comment]
> Agree re code structure. Is that something you want to tackle?

That would be fun but looking at my too-busy life I'm not sure that this will be reasonable for me, so feel free to do it (it's very possible that I'll end up doing it anyway).

> We should also keep an eye on performance. From just playing with the tests I noticed a very minimal slow down with these changes, but worth it for proper parsing.

Do you have experience in including that into CI/tests? That would allow us to avoid loosing sigh of this.

File: hamlpy/attribute_dict_parser.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        self.s = data.lstrip()
+        self.length=len(self.s)
+        # Index of current character being read
+        self.ptr=1
+
+    def consume_whitespace(self, include_newlines=False):
+        """Moves the pointer to the next non-whitespace character"""
+        whitespace = (' ', '\t', '\r', '\n') if include_newlines else (' ', '\t')
+
+        while self.ptr<self.length and self.s[self.ptr] in whitespace:
```
Comment (rowanseymour) [inline hamlpy/attribute_dict_parser.py]: I think on line 111 the reason for `self.length-1` is that it's parsing up to but not including the final `}`

**Comment by rowanseymour** [comment]
Not sure about performance tests in CI, but I'm gonna look at making it easier to run repeated timed tests on the set of hamlpy/html files

**Comment by rowanseymour** [comment]
@Psycojoker I've merged and am now gonna have a look at improving the tests. Can you get your PEP changes into a PR or just push them?

**Comment by Psycojoker** [comment]
@rowanseymour done.

Do you think we are ready for a release? You weren't as confident as for the previous PR iirc

**Comment by rowanseymour** [comment]
@Psycojoker I'd like to finish working on the tests first - see if that brings up any issues

**Pull request merged**